### PR TITLE
Cache manager

### PR DESCRIPTION
**Pull request by JerryImMouse**

### CODE REVIEW

File: Hypercube.Shared/Resources/Caching/Resource/BaseResource.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    public abstract void Load(ResourcePath path, DependenciesContainer container);
+
+    public virtual void Reload(ResourcePath path, DependenciesContainer container)
+    {
+        
+    }
+    
+    public ResourcePath? FallbackPath { get; }
+
+    public virtual void Dispose()
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Caching/Resource/BaseResource.cs:16]:
```C#
// Implement IDisposable.
// Do not make this method virtual.
// A derived class should not be able to override this method.
```

https://learn.microsoft.com/ru-ru/dotnet/api/system.idisposable?view=net-8.0

File: Hypercube.Shared/Resources/Caching/Resource/BaseResource.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+﻿using Hypercube.Shared.Dependency;
+
+namespace Hypercube.Shared.Resources.Caching.Resource;
+
+public abstract class BaseResource : IDisposable
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Caching/Resource/BaseResource.cs:5]: Just `Resource`

File: Hypercube.Shared/Resources/Caching/Resource/BaseResource.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public abstract void Load(ResourcePath path, DependenciesContainer container);
+
+    public virtual void Reload(ResourcePath path, DependenciesContainer container)
+    {
+        
+    }
+    
+    public ResourcePath? FallbackPath { get; }
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Caching/Resource/BaseResource.cs:14]: Properties declarated under construcotrs and methods

File: Hypercube.Shared/Resources/Caching/ICacheManager.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using Hypercube.Shared.Resources.Caching.Resource;
+
+namespace Hypercube.Shared.Resources.Caching;
+
+public interface ICacheManager
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Caching/ICacheManager.cs:6]: Better name `IResourceCacher`

File: Hypercube.Client/Resources/Caching/TextureResource.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+﻿using Hypercube.Client.Graphics.Texturing;
+using Hypercube.Client.Graphics.Texturing.TextureSettings;
+using Hypercube.Shared.Dependency;
+using Hypercube.Shared.Resources;
+using Hypercube.Shared.Resources.Caching.Resource;
+using Hypercube.Shared.Resources.Manager;
+
+namespace Hypercube.Client.Resources.Caching;
+
+public class TextureResource : BaseResource
```
Comment (Tornado-Technology) [inline Hypercube.Client/Resources/Caching/TextureResource.cs:10]: Make class `sealed`

File: Hypercube.Client/Resources/Caching/TextureResource.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    public override void Load(ResourcePath path, DependenciesContainer container)
+    {
+        var textureManager = container.Resolve<ITextureManager>();
+        var handle = textureManager.GetTextureHandle(path, new Texture2DCreationSettings());
+        Texture = handle;
+    }
+
+    public override void Dispose()
+    {
+        base.Dispose();
```
Comment (Tornado-Technology) [inline Hypercube.Client/Resources/Caching/TextureResource.cs:24]: `ITextureHandle` needed to dispose

File: Hypercube.Client/Resources/Caching/CacheManager.cs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+using Hypercube.Shared.Resources.Caching.Resource;
+using Hypercube.Shared.Resources.Manager;
+
+namespace Hypercube.Client.Resources.Caching;
+
+public partial class CacheManager : ICacheManager
+{
+    [Dependency] private readonly IResourceManager _resourceManager = default!;
+    
+    private Dictionary<Type, Dictionary<ResourcePath, BaseResource>> _cachedResources =
```
Comment (Tornado-Technology) [inline Hypercube.Client/Resources/Caching/CacheManager.cs:15]: Just `new()`

File: Hypercube.Client/Resources/Caching/CacheManager.cs [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+    private DependenciesContainer _container = default!;
+
+    private readonly Logger _logger = LoggingManager.GetLogger("cache");
+    
+    #region PublicAPI
+
+    public CacheManager()
+    {
+        _container = DependencyManager.GetContainer();
```
Comment (Tornado-Technology) [inline Hypercube.Client/Resources/Caching/CacheManager.cs:26]: You don't need construcot for this.

File: Hypercube.Client/Resources/Caching/CacheManager.cs
Comment (Tornado-Technology) [inline Hypercube.Client/Resources/Caching/CacheManager.cs:1]: Remove regions please

File: Hypercube.Client/Resources/Caching/CacheManager.cs [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            cache.Load(path, _container);
+            typeDict[path] = cache;
+            return (T) cache;
+        }
+        catch (Exception ex)
+        {
+            if (useFallback && cache.FallbackPath is not null)
+                return GetResource<T>(cache.FallbackPath.Value, false);
+            
+            _logger.Error($"Exception while loading resource {ex.Message}, Stack Trace: {ex.StackTrace}");
```
Comment (Tornado-Technology) [inline Hypercube.Client/Resources/Caching/CacheManager.cs:49]: It's `Fatal`, not `Error`

File: Hypercube.Client/Resources/Caching/CacheManager.cs [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    public void CacheResource<T>(ResourcePath path, T resource) where T : BaseResource, new()
+    {
+        GetTypeDict<T>()[path] = resource;
+    }
+
+    public T GetFallback<T>() where T : BaseResource, new()
+    {
+        throw new NotImplementedException();
```
Comment (Tornado-Technology) [inline Hypercube.Client/Resources/Caching/CacheManager.cs:92]: Do implementation

File: Hypercube.Client/Resources/Caching/CacheManager.Preload.cs [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+public partial class CacheManager
+{
+    public void PreloadTextures()
+    {
+        var logger = LoggingManager.GetLogger("cache.preload");
+        
+        PreloadTextures(logger);
+    }
+    private void PreloadTextures(Logger logger)
```
Comment (Tornado-Technology) [inline Hypercube.Client/Resources/Caching/CacheManager.Preload.cs:15]: Forgot the offset between methods

File: Hypercube.Client/Resources/Caching/CacheManager.Preload.cs [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+using Hypercube.Shared.Dependency;
+using Hypercube.Shared.Logging;
+
+namespace Hypercube.Client.Resources.Caching;
+
+public partial class CacheManager
+{
+    public void PreloadTextures()
+    {
+        var logger = LoggingManager.GetLogger("cache.preload");
```
Comment (Tornado-Technology) [inline Hypercube.Client/Resources/Caching/CacheManager.Preload.cs:11]: Decalare second logger, with name like `_loggerPreload`

**Review by Tornado-Technology — CHANGES_REQUESTED** [review]
A little review, on pointing out some bugs

File: Hypercube.Shared/Resources/Caching/Resource/BaseResource.cs (hunk 1 above)
Comment (JerryImMouse) [reply Hypercube.Shared/Resources/Caching/Resource/BaseResource.cs]: Done

**Review by Tornado-Technology — APPROVED** [review]

**Pull request merged**